File: monitoring-server/src/main/java/com/example/monitoringserver/service/ProcessService.java
package com.example.monitoringserver.service;

import com.example.monitoringserver.model.MessageDetail;
import com.example.monitoringserver.model.MessageId;
import com.example.monitoringserver.model.Process;
import com.example.monitoringserver.model.ProcessId;

import java.util.List;
import java.util.Optional;

public interface ProcessService {

	List<Process> getProcesses();

	Optional<Process> getProcess(ProcessId processId);

	Optional<MessageDetail> getFailedMessageDetail(ProcessId processId, MessageId messageId);

	Optional<MessageDetail> getPendingMessageDetail(ProcessId processId, MessageId messageId);
}

File: monitoring-server-ui/src/app/processes/services/process.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs/Observable';

import {Process} from '../models/process';
import {MessageDetail} from '../models/message-detail';

@Injectable()
export class ProcessService {
  private processesUrl = '/api/processes';

  constructor(private http: HttpClient) {
  }

  getProcesses(): Observable<Process[]> {
    return this.http.get<Process[]>(this.processesUrl);
  }

  getProcess(id: string): Observable<Process> {
    return this.http.get<Process>(`${this.processesUrl}/${id}`);
  }

  getFailedMessage(processId: string, messageId: string): Observable<MessageDetail> {
    return this.http.get<MessageDetail>(`${this.processesUrl}/${processId}/failed-messages/${messageId}`);
  }

  getPendingMessage(processId: string, messageId: string): Observable<MessageDetail> {
    return this.http.get<MessageDetail>(`${this.processesUrl}/${processId}/pending-messages/${messageId}`);
  }

}
